### PR TITLE
Fixes #1884 Create domain object for Module

### LINKED ISSUE #1884: Create domain object for Module
To start on modularization we need a domain object to represent a collection of building blocks that will be called a 'Module'.

The module has at most one building block of each type.

### PR DESCRIPTION
**Pull request by rwmcintosh**

### CODE REVIEW

File: src/OSPSuite.Core/Domain/BaseModule.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+using System.Collections.Generic;
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Core.Domain.Services;
+using OSPSuite.Utility.Extensions;
+
+namespace OSPSuite.Core.Domain
+{
+   public abstract class BaseModule : ObjectBase
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/BaseModule.cs:8]: Module? Why BaseModule

File: src/OSPSuite.Core/Domain/BaseModule.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+         EventBlock = cloneManager.Clone(sourceModule.EventBlock);
+
+         sourceModule.MoleculeStartValueBlockCollection.Each(x => MoleculeStartValueBlockCollection.Add(cloneManager.Clone<IMoleculeStartValuesBuildingBlock>(x)));
+         sourceModule.ParametersStartValueBlockCollection.Each(x => ParametersStartValueBlockCollection.Add(cloneManager.Clone(x)));
+      }
+   }
+
+   public class ExtensionModule : BaseModule
+   {
+      public override bool UserEditable => true;
```
[the remark this answers is not in the document]
Comment (rwmcintosh) [reply src/OSPSuite.Core/Domain/BaseModule.cs]: Sure can.

File: src/OSPSuite.Core/Domain/Module.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+         UserEditable = userEditable;
+      }
+      
+      public IMoleculeBuildingBlock MoleculeBlock { set; get; }
+      public IReactionBuildingBlock ReactionBlock { set; get; }
+      public IPassiveTransportBuildingBlock PassiveTransport { set; get; }
+      public ISpatialStructure SpatialStructure { set; get; }
+      public IObserverBuildingBlock ObserverBlock { set; get; }
+      public IEventGroupBuildingBlock EventBlock { set; get; }
+      public IList<IMoleculeStartValuesBuildingBlock> MoleculeStartValueBlockCollection { set; get; } = new List<IMoleculeStartValuesBuildingBlock>();
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Module.cs:26]: do not expose your internal Lists. this should be an IREADOnly

File: src/OSPSuite.Core/Domain/Module.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            return;
+
+         MoleculeBlock = cloneManager.Clone(sourceModule.MoleculeBlock);
+         ReactionBlock = cloneManager.Clone(sourceModule.ReactionBlock);
+         PassiveTransport = cloneManager.Clone(sourceModule.PassiveTransport);
+         SpatialStructure = cloneManager.Clone(sourceModule.SpatialStructure);
+         ObserverBlock = cloneManager.Clone(sourceModule.ObserverBlock);
+         EventBlock = cloneManager.Clone(sourceModule.EventBlock);
+
+         sourceModule.MoleculeStartValueBlockCollection.Each(x => MoleculeStartValueBlockCollection.Add(cloneManager.Clone<IMoleculeStartValuesBuildingBlock>(x)));
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Module.cs:45]: why do you need to specify the type here? Thsi is not consistent with wha'tsa bove

File: src/OSPSuite.Core/Domain/Module.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      public override void UpdatePropertiesFrom(IUpdatable source, ICloneManager cloneManager)
+      {
+         base.UpdatePropertiesFrom(source, cloneManager);
+         
+         if (!(source is Module sourceModule))
+            return;
+
+         MoleculeBlock = cloneManager.Clone(sourceModule.MoleculeBlock);
+         ReactionBlock = cloneManager.Clone(sourceModule.ReactionBlock);
+         PassiveTransport = cloneManager.Clone(sourceModule.PassiveTransport);
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/Module.cs:40]: verify in tests that htis works if one entry is null. Probalby does

Comment (msevestre) [reply src/OSPSuite.Core/Domain/Module.cs:40]: but this means you cannot pass a fake of the clone manager

**Review by msevestre — CHANGES_REQUESTED** [review]

**Review by msevestre — APPROVED** [review]

**Pull request merged**